techdemo.js

```diff
@@ -0,0 +1,323 @@
+
+    square.set('left', (lastX + mouse.x) / 2);
+    square.set('top', (lastY + mouse.y) / 2);
+
+    canvas.renderAll();
+}
+
+function DrawRectMouseUp(e) {
+    if (started) {
+        started = false;
+    }
+    var square = canvas.getActiveObject();
+    canvas.discardActiveObject();
+    square.setCoords();
+    canvas.renderAll();
+
+    //    selected_Object = undefined;
+}
+
+
+
+//  Draw Line events
+
+function DrawLineMouseDown(e) {
+    var mouse = canvas.getPointer(e.e);
+    started = true;
+    lastX = mouse.x;
+    lastY = mouse.y;
+
+    var line = new fabric.Line([lastX, lastY, lastX, lastY], {
+        stroke: 'red',
+        strokeWidth: 3,
+        lockScalingX: true,
+        lockScalingY: true,
+        lockRotation: true,
+        hasBorders: false,
+        hasControls: false,
+        perPixelTargetFind: true,
+        fill: 'red'
+    });
+    line.lockScalingX = line.lockScalingY = true;
+
+    var circle1 = new fabric.Circle({
+        left: lastX,
+        top: lastY,
+        lockScalingX: true,
+        lockScalingY: true,
+        lockRotation: true,
+        hasBorders: false,
+        hasControls: false,
+        radius: 3,
+        strokeWidth: 3,
+        stroke: 'red',
+        fill: 'red'
+    });
+    var circle2 = new fabric.Circle({
+        left: lastX,
+        top: lastY,
+        lockScalingX: true,
+        lockScalingY: true,
+        lockRotation: true,
+        hasBorders: false,
+        hasControls: false,
+        radius: 3,
+        strokeWidth: 3,
+        stroke: 'red',
+        fill: 'red'
+    });
+
+    circle1.line = circle2.line = line;
+    line.circle1 = circle1;
+    line.circle2 = circle2;
+
+    canvas.add(line);
+
+    canvas.add(circle1);
+    canvas.add(circle2);
+
+    circle1.bringToFront();
+    circle2.bringToFront();
+
+    canvas.setActiveObject(line);
+    canvas.renderAll();
+}
+
+function DrawLineMouseMove(e) {
+    if (!started) {
+        return false;
+    }
+    var mouse = canvas.getPointer(e.e);
+
+    var line = canvas.getActiveObject();
+    line.set('x2', mouse.x).set('y2', mouse.y);
+    line.setCoords();
+
+    var cir1 = line.get('circle1');
+    var cir2 = line.get('circle2');
+    cir1.set('left', mouse.x).set('top', mouse.y);
+    cir1.setCoords();
+
+    canvas.renderAll();
+}
+
+function DrawLineMouseUp(e) {
+    if (started) {
+        started = false;
+    }
+    var line = canvas.getActiveObject();
+    canvas.discardActiveObject();
+    line.setCoords();
+    canvas.renderAll();
+
+    selected_Object = undefined;
+}
+
+
+function ObjectMoving(e) {
+    var p = e.target;
+
+    // move line
+    if (p.circle1) {
+        var oldCenterX = (p.x1 + p.x2) / 2;
+        var oldCenterY = (p.y1 + p.y2) / 2;
+
+        var deltaX = p.left - oldCenterX;
+        var deltaY = p.top - oldCenterY;
+
+        p.circle1 && p.circle1.set({
+            'left': p.x1 + deltaX,
+                'top': p.y1 + deltaY
+        }).setCoords();
```

Comment:
__[jshint]__
*  Expected an assignment or function call and instead saw an expression.


<!-- Comment by [SideCI](https://sideci.com) -->